### DISCUSSION TITLE
Enable Webmentions

### DISCUSSION THREAD

**Discussion by Cheukting** [forem/forem#15394] [Feature Requests]
**Is your feature request related to a problem? Please describe.**

My website can receive webmentions, and I keep syndicating my blog post to dev.to. If the reactions that I received on dev.to can be reflected on my website that would be great.

**Describe the solution you'd like**

dev.to enable webmentions and pingback to my original blog post on my website whenever my post on dev.to received a reaction.

**Describe alternatives you've considered**

Not sure if there are any.

**Additional context**

**Comment by rhymes** [comment]
Thanks for the issue! We'll take your request into consideration and follow up if we decide to tackle this issue.

To our amazing contributors: issues labeled `type: bug` are always up for grabs, but for feature requests, please wait until we add a `ready for dev` before starting to work on it.

To claim an issue to work on, please leave a comment. If you've claimed the issue and need help, please ping @thepracticaldev/oss and we will follow up within 3 business days.

For full info on how to contribute, please check out our [contributors guide](https://github.com/thepracticaldev/dev.to/blob/master/CONTRIBUTING.md).

**Comment by rhymes** [comment]
Hi @Cheukting, TIL about Webmentions!

They remind me of the "pingbacks" of olden days.

**Comment by citizen428** [comment]
I very much like this idea. I looked into web mentions the other day and I think they fit very well with our overall goal of openness.

**Comment by muety** [comment]
I totally support the request to implement [Webmentions](https://www.w3.org/TR/webmention/) and / or [Pingback](https://www.w3.org/TR/webmention/#bib-Pingback) integration! 🙂 Since very community- and open-source focused, dev.to is a perfect candidate for pushing those open web standards / protocols.

**Comment by benhalpern** [comment]
I have been somewhat non-committal mostly because I haven't quite wrapped my head around how exactly it might go down....

But since we agree in principal with big time all this I'm going to mark this as `ready for dev` and maybe we can start taking baby steps in getting this all properly integrated.

**Comment by aaronpk** [comment]
I took a stab at outlining how this might work. Here's the writeup, including details on the (minimal) markup additions needed to Dev.to comment pages:

https://indieweb.org/Webmention-brainstorming#receiving_webmentions_for_syndicated_copies

**Comment by citizen428** [comment]
Thanks for this @aaronpk, this will make a nice basis for discussions! ❤️

**Comment by eyewritecode** [comment]
Echoing what @rhymes said, i also didn't know about Webmentions and would enjoy owning this issue if no one is looking into this already! cc: @citizen428 @benhalpern ?

**Comment by rhymes** [comment]
@eyewritecode I assigned you to the issue as it's open for development. It'd be great to come up with a mini plan on how to tackle this so we can discuss it before you set out to implement it. How does that sound?

**Comment by eyewritecode** [comment]
Cool, thanks @rhymes! 
Yeah i'll definitely share my approach once i get the hang of how it works and how it fits in with DEV.

**Comment by aaronpk** [comment]
Feel free to join the dev channel on IndieWeb chat if you'd like to bounce ideas off of people for this! https://indieweb.org/discuss

Also https://webmention.rocks will be a good resource for you while testing!

**Comment by eyewritecode** [comment]
Thanks @aaronpk, your blog posts have been of a great help so far :)

**Comment by muety** [comment]
Hey guys, any progress on this? Can't wait to have Webmention support! :slightly_smiling_face:

**Comment by eyewritecode** [comment]
Thanks for bringing this back on my radar @muety 

I finally had time to look into this and below is how i'd approach this. 
FYI, This is more of a combination of what Aaron had suggested [here](https://indieweb.org/Webmention-brainstorming#receiving_webmentions_for_syndicated_copies) and how i think it fits in DEV codebase..

For  webmentions to work, we basically need two things:

- Original link that “accepts” Webmentions
- “Copy” of the original post(I.e post on DEV)

Since DEV allows users to add a canonical URL to their posts, We can either:
 - Assume that this URL accepts web mentions(easiest solution)
 - Offer another option where users specifically provide a webmention URL  
 - Provide a checkbox for users who want to receive webmentions on their canonical URL

The backend implementation is similar for any of the above options and that's basically having a background job that does a few things:

- Check whether the provided canonical URL accepts webmentions(i.e Parse that URL for “rel=‘webmention’”)
 - If true, schedule a POST request to their webmention URL that would update their original URL with new DEV's reactions(comments, hearts, etc...)
I think this should have the same priority as the worker for counting a posts views 

Also, Aaron went ahead and suggested Microformats needed on the post's comment section https://gist.github.com/aaronpk/85e4adc582aeb6add7876b9f5a006e42/revisions which is awesome :)

I'm probably missing on some stuff but i think this should be a start for discussion on some edge cases! what do you think @rhymes

**Comment by muety** [comment]
Very good thoughts! Although I'm not into the code base of dev.to at all, @eyewritecode's approach makes a lot of sense to me.

In addition, I was wondering if dev.to's support for webmentions might even be extended beyond sending webmention requests for comments on syndicated posts. To my understanding (please correct me!), dev.to could also ...

* ... crawl any (not only syndications) new blog post on dev.to for external links and attempt to send a webmention request to each of them. This way, the author of a post on some private blog can be notified when their post is referenced in a recent dev.to article
* ... offer the ability to dev.to authors to enable their post for webmentions. That is, every article on dev.to would optionally include the `rel="webmention"` link. This way, the dev.to author A could be notified (through dev.to) when some author B on an external blog references A's article. Given that B's blog has webmention support as well, it would send a webmention callback to dev.to's API, which, in turn, somehow notifies A. 

Does this align with the idea of webmentions? And if yes, do you consider those points an option for dev.to?

**Comment by citizen428** [comment]
Hi @eyewritecode, thanks for the summary! The overall approach seems sound, I advocate for getting a PR up for a minimal implementation of this and iterate from there.

**Comment by Seirdy** [comment]
On the topic for sending Webmentions for syndicated copies:

If a post on Dev.to is a syndicated copy of another post on the author's own Webmention-supporting website, it would be really great if the JSON-LD metadata for Dev.to posts changed its `author` fields to match the `h-card` fields on the primary location's `h-card`.

For a start, changing `author.url` to match their `h-card`'s `url` would be nice.

**Comment by cmgorton** [comment]
Hello all! Thank you for opening this request. There has been a lot of good discussion here but has since gone stale. 
The Forem core team now uses an internal RFC ("request for comments") process to assess and prioritize new features. This process is intended to provide a consistent and standardized path for new changes to enter the Forem ecosystem. If you'd like to propose a new feature like this one, please visit [forem.dev](https://forem.dev/) to start a discussion around the feature! 🙏 I think it would be great if we could continue the discussion from this issue over there and get more visibility from our Product team on this. Thanks!

**Comment by Seirdy** [comment]
On Mon, Mar 29, 2021 at 01:02:46PM -0700, Christina wrote:
>If you'd like to propose a new feature like this one, please visit 
>[forem.dev](https://forem.dev/)...continue the discussion from this 
>issue over there

Sure, but I'm not generally a fan of creating a new account for these 
things.

-- 
/Seirdy

**Comment by cmgorton** [comment]
Totally understand. Before you decide whether you would like to create an account you can read more on why we made the decision to take feature requests off of Github and on forem.dev in this post.  https://forem.dev/foremteam/internal-rfc-process-and-forem-dev-discussions-3gl4 
You can also see some examples of other feature requests in these round ups. 
https://forem.dev/coffeecraftcode/let-s-discuss-a-round-up-of-feature-requests-5clh

**Comment by siakaramalegos** [comment]
Hi, I wanted to connect this to the analogous issue in Bridgy https://github.com/snarfed/bridgy/issues/874 - At least in 2019 Dev was doing an API, but it wasn't document yet for us to be able to ping it to get the necessary data.

**Comment by siakaramalegos** [comment]
I added a post on Forem too: https://forem.dev/thegreengreek/enable-webmentions-o0c
